dask/array/tests/test_cupy.py

```diff
@@ -78,3 +103,490 @@ def test_sizeof(dtype):
+
+    with pytest.raises(ValueError):
+        da.diagonal(v, axis1=-2)
+
+    # Negative axis.
+    assert_eq(da.diagonal(v, axis1=-1), np.diagonal(v, axis1=-1))
+    assert_eq(da.diagonal(v, offset=1, axis1=-1), np.diagonal(v, offset=1, axis1=-1))
+
+    # Heterogenous chunks.
+    v = cupy.arange(2 * 3 * 4 * 5 * 6).reshape((2, 3, 4, 5, 6))
+    v = da.from_array(v, chunks=(1, (1, 2), (1, 2, 1), (2, 1, 2), (5, 1)), asarray=False)
+
+    assert_eq(da.diagonal(v), np.diagonal(v))
+    assert_eq(da.diagonal(v, offset=2, axis1=3, axis2=1),
+              np.diagonal(v, offset=2, axis1=3, axis2=1))
+
+    assert_eq(da.diagonal(v, offset=-2, axis1=3, axis2=1),
+              np.diagonal(v, offset=-2, axis1=3, axis2=1))
+
+    assert_eq(da.diagonal(v, offset=-2, axis1=3, axis2=4),
+              np.diagonal(v, offset=-2, axis1=3, axis2=4))
+
+    assert_eq(da.diagonal(v, 1), np.diagonal(v, 1))
+    assert_eq(da.diagonal(v, -1), np.diagonal(v, -1))
+    # Positional arguments
+    assert_eq(da.diagonal(v, 1, 2, 1), np.diagonal(v, 1, 2, 1))
+
+
+@pytest.mark.xfail(reason="no shape argument support *_like functions on CuPy yet")
+@pytest.mark.skipif(np.__version__ < '1.17', reason='no shape argument for *_like functions')
+@pytest.mark.skipif(missing_arrfunc_cond, reason=missing_arrfunc_reason)
+def test_tril_triu():
+    A = cupy.random.randn(20, 20)
+    for chk in [5, 4]:
+        dA = da.from_array(A, (chk, chk), asarray=False)
+
+        assert np.allclose(da.triu(dA).compute(), np.triu(A))
+        assert np.allclose(da.tril(dA).compute(), np.tril(A))
+
+        for k in [-25, -20, -19, -15, -14, -9, -8, -6, -5, -1,
+                  1, 4, 5, 6, 8, 10, 11, 15, 16, 19, 20, 21]:
+            assert np.allclose(da.triu(dA, k).compute(), np.triu(A, k))
+            assert np.allclose(da.tril(dA, k).compute(), np.tril(A, k))
+
+
+@pytest.mark.xfail(reason="no shape argument support *_like functions on CuPy yet")
+@pytest.mark.skipif(np.__version__ < '1.17', reason='no shape argument for *_like functions')
+@pytest.mark.skipif(missing_arrfunc_cond, reason=missing_arrfunc_reason)
+def test_tril_triu_non_square_arrays():
+    A = cupy.random.randint(0, 11, (30, 35))
+    dA = da.from_array(A, chunks=(5, 5), asarray=False)
+    assert_eq(da.triu(dA), np.triu(A))
+    assert_eq(da.tril(dA), np.tril(A))
+
+
+@pytest.mark.skipif(missing_arrfunc_cond, reason=missing_arrfunc_reason)
+def test_apply_gufunc_axis():
+    def mydiff(x):
+        return np.diff(x)
+
+    a = cupy.random.randn(3, 6, 4)
+    da_ = da.from_array(a, chunks=2, asarray=False)
+
+    m = np.diff(a, axis=1)
+    dm = apply_gufunc(mydiff, "(i)->(i)", da_, axis=1, output_sizes={'i': 5},
+                      allow_rechunk=True)
+    assert_eq(m, dm)
+
+
+def test_overlap_internal():
+    x = cupy.arange(64).reshape((8, 8))
+    d = da.from_array(x, chunks=(4, 4), asarray=False)
+
+    g = da.overlap.overlap_internal(d, {0: 2, 1: 1})
+    result = g.compute(scheduler='sync')
```

Comment:
No, we don't. Again a copy from pre-existing test, removing that now.

Reply:
In the future (after this PR) we might consider a pytest fixture to convert numpy arrays to other array types that we could insert into other tests

```python
# dask/array/tests/test_linalg.py
def test_svd(normalize):
    x = np.random.random(...)
    x = normalize(x)  # maybe convert x to cupy or sparse array

    ...
```

This might be a more composable way to add things like cupy tests in the future.

(but again, I would wait until some future PR)

Reply:
Yes, I was thinking of some solution, but I didn't want to insert `cupy` anywhere but `test_cupy.py` for the time being. Indeed, long-term it would be much better than copying tests like I'm doing here.